.github/workflows/release.yml

```diff
@@ -1,68 +1,91 @@
 name: release
+
+concurrency:
+  group: ${{ github.workflow }}-${{ github.head_ref || github.ref }}
+  cancel-in-progress: true
+
 on:
   push:
     tags:
-      - "v[0-9]+.[0-9]+.*"
+      - 'v[0-9]+.[0-9]+.*'
+
+env:
+  CARGO_TERM_COLOR: always
+  REPO_SLUG: "deislabs/runwasi"
+
 jobs:
-  build:
+  hub:
     runs-on: ubuntu-latest
     steps:
-      - uses: actions/checkout@v3
-      - uses: actions/cache@v3
+      - name: Checkout
+        uses: actions/checkout@v3
+
+      - name: Set up Docker Buildx
+        uses: docker/setup-buildx-action@v2
+
+      - name: Docker meta
+        id: meta
+        uses: docker/metadata-action@v4
         with:
-          path: |
-            ~/.cargo/registry
-            ~/.cargo/git
-            target
-          key: ${{ runner.os }}-cargo-${{ hashFiles('**/Cargo.lock') }}
-      - name: Build
-        run: cargo build --verbose
-      - name: Run tests
-        run: cargo test --verbose
+          images: ${{ env.REPO_SLUG }}
+          ### frontend versioning
+          ### on semver tag:
+          # deislabs/runwasi:1.2.3
+          # deislabs/runwasi:1.2
+          # deislabs/runwasi:1
+          # deislabs/runwasi:latest
+          ### on pre-release tag:
+          # deislabs/runwasi:1.1.0-rc.1
+          ### on push default branch (main):
+          # deislabs/runwasi:main
+          tags: |
+            type=semver,pattern={{version}}
+            type=semver,pattern={{major}}.{{minor}}
+            type=semver,pattern={{major}}
+            type=ref,event=branch
+            type=raw,value=latest
+            type=ref,event=pr
+          bake-target: meta-helper
+          flavor: |
+            latest=false
+      - name: Login to DockerHub
+        if: ${{ github.event_name != 'pull_request' }}
+        uses: docker/login-action@v2
+        with:
+          username: ${{ secrets.DOCKER_HUB_USER }}
+          password: ${{ secrets.DOCKER_HUB_TOKEN }}
+      - name: Build and push
+        uses: docker/bake-action@v2
+        with:
+          files: |
+            ./docker-bake.hcl
+            ${{ steps.meta.outputs.bake-file }}
+          targets: image-cross
+          push: ${{ github.event_name != 'pull_request' }}
+
   release:
     permissions:
       contents: write
-    needs: build
     if: startsWith(github.ref, 'refs/tags/v')
+    needs: hub
```

Comment:
How come this is changing to use dockerhub to store artifacts?
I'd rather keep this inside github and not add the extra dependency.

Reply:
1. **The hub section in workflow should be keep** for docker-desktop could integration it. (In our work before we only deliver docker image to docker team. The artifacts is inessentials  although we provide it, too). So my first through is reduce compile time and provide cross platform artifacts.
2. But I still decoupling build docker image and release artifact steps at aba24c7a6f38dc6d50ca719535b4587cd112f42c. The side effect is we could provide amd version artifact only. If you could provide an arm runner and consider to add arm version artifact in the future maybe you could ref to [our workflow](https://github.com/second-state/runwasi/blob/736290ec1b962430199a0bb1a0c1aa81b85baa3a/.github/workflows/release.yml#L76-L114).